src/context.rs

```diff
@@ -33,76 +42,83 @@ pub struct OutstandingRequests {
     youngest: Option<Id>,
 }
 
-pub struct Context {
-    batch_counter: BatchNumber,
-    pub batches:
-        HashMap<BatchNumber, (BatchCount, Vec<serde_json::value::Value>, ResponsesCallback)>,
+pub struct ServerSettings {
+    pub root_path: String,
+    pub offset_encoding: OffsetEncoding,
+    pub preferred_offset_encoding: Option<OffsetEncoding>,
     pub capabilities: Option<ServerCapabilities>,
-    pub completion_items: Vec<CompletionItem>,
+    pub tx: Sender<ServerMessage>,
+}
+
+pub struct Context {
+    batch_count: BatchCount,
+    pub batch_sizes: HashMap<BatchNumber, HashMap<ServerName, usize>>,
+    pub batches: HashMap<
+        BatchNumber,
+        (
+            Vec<(ServerName, serde_json::value::Value)>,
+            ResponsesCallback,
+        ),
+    >,
+    pub code_lenses: HashMap<String, Vec<(ServerName, CodeLens)>>,
+    pub completion_items: Vec<(ServerName, CompletionItem)>,
```

Comment:
Have you given thought to moving these two into `ServerSettings`?
I guess that would be less convenient because of the way we forward it to the editor.